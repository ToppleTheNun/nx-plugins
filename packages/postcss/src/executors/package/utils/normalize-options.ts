import { normalizePath } from "@nrwl/devkit";
import {
  NormalizedPackageExecutorSchema,
  PackageExecutorSchema,
} from "../schema";

export const normalizePackageOptions = (
  options: PackageExecutorSchema,
  root: string,
  sourceRoot: string,
  projectRoot: string
): NormalizedPackageExecutorSchema => {
  return {
    ...options,
    root: normalizePath(root),
    sourceRoot: normalizePath(sourceRoot),
    projectRoot: normalizePath(projectRoot),
    main: normalizePath(options.main),
    outputPath: normalizePath(options.outputPath),
    watch: options.watch ?? false,
    outputFileName: options.outputFileName ?? "index.css",
  };
};
