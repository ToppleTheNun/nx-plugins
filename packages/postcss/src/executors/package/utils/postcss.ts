import { joinPathFragments, normalizePath } from "@nrwl/devkit";
import { ensureFile } from "fs-extra";
import { readFile, writeFile } from "fs/promises";
import postcss from "postcss";
import * as postcssrc from "postcss-load-config";
import { NormalizedPackageExecutorSchema } from "../schema";

export async function readAndCompileCss(
  options: NormalizedPackageExecutorSchema
) {
  const rawInputCss = await readFile(
    joinPathFragments(normalizePath(options.root), options.main),
    "utf-8"
  );

  const postcssrcConfig: postcssrc.ConfigContext = {
    cwd: normalizePath(options.projectRoot),
    env: "production",
    from: options.main,
    to: joinPathFragments(options.outputPath, options.outputFileName),
    map: options.sourceMap ?? { inline: true },
  };
  const { options: postcssOptions, plugins: postcssPlugins } = await postcssrc(
    postcssrcConfig,
    normalizePath(options.projectRoot)
  );
  const postcssResult = await postcss(postcssPlugins).process(
    rawInputCss,
    postcssOptions
  );

  const pathToOutputFile = joinPathFragments(
    normalizePath(options.root),
    options.outputPath,
    options.outputFileName
  );
  await ensureFile(pathToOutputFile);
  await writeFile(pathToOutputFile, postcssResult.css, "utf-8");
}
