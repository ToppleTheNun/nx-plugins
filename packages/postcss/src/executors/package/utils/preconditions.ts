import { ExecutorContext } from "@nrwl/devkit";
import { readCachedProjectGraph } from "@nrwl/workspace/src/core/project-graph";
import {
  calculateProjectDependencies,
  checkDependentProjectsHaveBeenBuilt,
} from "@nrwl/workspace/src/utilities/buildable-libs-utils";

export function requireRootAndSourceRoot(context: ExecutorContext) {
  const { sourceRoot, root } = context.workspace.projects[context.projectName];
  if (!sourceRoot) {
    throw new Error(`${context.projectName} does not have a sourceRoot.`);
  }
  if (!root) {
    throw new Error(`${context.projectName} does not have a root.`);
  }
  return { sourceRoot, root };
}

export function requireDependentProjectsToBeBuilt(context: ExecutorContext) {
  const projGraph = readCachedProjectGraph();
  const { dependencies } = calculateProjectDependencies(
    projGraph,
    context.root,
    context.projectName,
    context.targetName,
    context.configurationName
  );

  if (
    !checkDependentProjectsHaveBeenBuilt(
      context.root,
      context.projectName,
      context.targetName,
      dependencies
    )
  ) {
    throw new Error(
      `${context.projectName} requires dependencies to be built first`
    );
  }
}
