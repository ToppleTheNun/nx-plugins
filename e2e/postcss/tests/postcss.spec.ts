import { newProject } from "@nrwl/e2e/utils";
import {
  checkFilesExist,
  readFile,
  runNxCommandAsync,
  uniq,
} from "@nrwl/nx-plugin/testing";

describe("PostCSS", () => {
  beforeAll(() =>
    newProject({
      npmPackageName: "@topplethenun/nx-plugin-postcss",
      pluginDistPath: "dist/packages/postcss",
    })
  );

  it("creates postcss.config.js", async () => {
    const mylib = uniq("generator-configuration");
    await runNxCommandAsync(`generate @nrwl/workspace:lib ${mylib}`);
    await runNxCommandAsync(
      `generate @topplethenun/nx-plugin-postcss:configuration --project ${mylib}`
    );

    expect(() => {
      checkFilesExist(`libs/${mylib}/postcss.config.js`);
    }).not.toThrow();
  }, 500000);

  it("can build a CSS file", async () => {
    const mylib = uniq("executor-build");
    await runNxCommandAsync(
      `generate @topplethenun/nx-plugin-postcss:library ${mylib} --buildable`
    );
    await runNxCommandAsync(`run-many --target=build --projects=${mylib}`);
    expect(() => {
      checkFilesExist(`dist/libs/${mylib}/dist/index.css`);
    }).not.toThrow();
    const actualContents = readFile(`dist/libs/${mylib}/dist/index.css`);
    expect(actualContents).toMatchSnapshot();
  }, 500000);
});
