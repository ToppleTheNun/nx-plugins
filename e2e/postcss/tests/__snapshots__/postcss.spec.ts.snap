// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCSS can build a CSS file 1`] = `
"::-moz-placeholder {
  color: red;
}
:-ms-input-placeholder {
  color: red;
}
::placeholder {
  color: red;
}

/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIi4uLy4uLy4uLy4uL2xpYnMvZXhlY3V0b3ItYnVpbGQ3MjgwOTk1L3NyYy9pbmRleC5jc3MiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBQUE7RUFDRSxVQUFVO0FBQ1o7QUFGQTtFQUNFLFVBQVU7QUFDWjtBQUZBO0VBQ0UsVUFBVTtBQUNaIiwiZmlsZSI6ImluZGV4LmNzcyIsInNvdXJjZXNDb250ZW50IjpbIjo6cGxhY2Vob2xkZXIge1xuICBjb2xvcjogcmVkO1xufVxuIl19 */"
`;
